privval/src/server.rs

```diff
@@ -0,0 +1,467 @@
+            },
+            None => {
+                error!("[{}] no public key found", req.chain_id);
+                PubKeyResponse {
+                    pub_key: None,
+                    error: Some(get_invalid_chain_id_error(&req.chain_id)),
+                }
+            },
+        };
+        Ok(Response::new(resp.into()))
+    }
+    async fn sign_vote(
+        &self,
+        request: Request<RawSignVoteRequest>,
+    ) -> Result<Response<RawSignedVoteResponse>, Status> {
+        let req = SignVoteRequest::try_from(request.into_inner())
+            .map_err(|e| Status::invalid_argument(e.to_string()))?;
+        debug!(
+            "[{}] received a vote signing request: {:?}",
+            req.chain_id, req
+        );
+        self.check_max_height(&req.chain_id, req.vote.height)
+            .map_err(|_| Status::failed_precondition("max height exceeded"))?;
+        let mut providers = self.providers.lock().await;
+        let resp = if let Some((signer, state_provider)) = providers.get_mut(&req.chain_id) {
+            let new_state = (&req).into();
+            let signable_bytes = req
+                .vote
+                .to_signable_vec(req.chain_id.clone())
+                .map_err(|e| Status::internal(e.to_string()))?;
+            match sign_and_persist_state(
+                &req.chain_id,
+                signer,
+                state_provider,
+                new_state,
+                signable_bytes,
+            )
+            .await
+            {
+                Ok(signature) => {
+                    let mut vote = req.vote.clone();
+                    vote.signature = Some(signature);
+                    SignedVoteResponse {
+                        vote: Some(vote),
+                        error: None,
+                    }
+                },
+                Err(err) => SignedVoteResponse {
+                    vote: None,
+                    error: Some(err),
+                },
+            }
+        } else {
+            error!("[{}] no signer found", req.chain_id);
+            SignedVoteResponse {
+                vote: None,
+                error: Some(get_invalid_chain_id_error(&req.chain_id)),
+            }
+        };
+        Ok(Response::new(resp.into()))
+    }
+    async fn sign_proposal(
+        &self,
+        request: Request<RawSignProposalRequest>,
+    ) -> Result<Response<RawSignedProposalResponse>, Status> {
+        let req = SignProposalRequest::try_from(request.into_inner())
+            .map_err(|e| Status::invalid_argument(e.to_string()))?;
+        debug!(
+            "[{}] received a proposal signing request: {:?}",
+            req.chain_id, req
+        );
+        self.check_max_height(&req.chain_id, req.proposal.height)
+            .map_err(|_| Status::failed_precondition("max height exceeded"))?;
+        let mut providers = self.providers.lock().await;
+        let resp = if let Some((signer, state_provider)) = providers.get_mut(&req.chain_id) {
```

Comment:
Similar comment to [this one](https://github.com/informalsystems/tendermint-rs/pull/1152/files#r927788402) here as well